Pegasus/include/particlecontacts.hpp

```diff
@@ -1,72 +1,167 @@
-#ifndef PEGAS_PARTICLE_CONTACTS_HPP
-#define PEGAS_PARTICLE_CONTACTS_HPP
+#ifndef PEGASUS_PARTICLE_CONTACTS_HPP
+#define PEGASUS_PARTICLE_CONTACTS_HPP
 
+#include "Pegasus/include/geometry.hpp"
+#include "Pegasus/include/mechanics.hpp"
 #include "Pegasus/include/particle.hpp"
+
 #include <vector>
 
-namespace pegas {
+namespace pegasus {
+
 class ParticleContact {
 public:
-    using Ptr = std::shared_ptr<ParticleContact>;
-
-    ParticleContact(Particle::Ptr const& a, Particle::Ptr const& b,
-        real const restitution, Vector3 const& contactNormal,
-        real const penetration);
-
-    void resolve(real const duration);
+    ParticleContact(Particle & a, Particle * b,
+                    double restitution, Vector3 const& contactNormal, double penetration);
 
-    real calculateSeparatingVelocity() const;
+    void resolve(double duration) const;
+    double calculateSeparatingVelocity() const;
 
 private:
-    Particle::Ptr mA;
-    Particle::Ptr mB;
-    real mRestitution;
-    Vector3 const mContactNormal;
-    real mPenetration;
+    Particle * mA;
+    Particle * mB;
+    double mRestitution;
+    Vector3 mContactNormal;
+    double mPenetration;
 
-    void resolveVelocity(real const duration);
-
-    void resolveInterpenetration(real const duration);
+private:
+    void resolveVelocity(double duration) const;
+    void resolveInterpenetration(double duration) const;
 };
 
-using ParticleContactsArray = std::vector<ParticleContact::Ptr>;
+using ParticleContacts = std::vector<ParticleContact>;
 
 class ParticleContactResolver {
 public:
-    ParticleContactResolver(unsigned int const iterations = 0);
-
-    void setIterations(unsigned int const iterations);
+    explicit ParticleContactResolver(uint32_t iterations = 0);
 
-    void resolveContacts(ParticleContactsArray& contacts,
-        real const duration);
+    void setIterations(uint32_t iterations);
+    void resolveContacts(ParticleContacts & contacts, double duration);
 
 private:
-    unsigned int mIterations;
-    unsigned int mIterationsUsed;
+    uint32_t mIterations;
+    uint32_t mIterationsUsed;
 };
 
 class ParticleContactGenerator {
 public:
-    using Ptr = std::shared_ptr<ParticleContactGenerator>;
-    using Contacts = std::vector<pegas::ParticleContact::Ptr>;
-
     virtual ~ParticleContactGenerator();
-    virtual unsigned int addContact(Contacts& contacts,
-        unsigned int const limit) const = 0;
+    virtual uint32_t addContact(ParticleContacts & contacts, uint32_t limit) const = 0;
 };
 
+template < typename Particles >
 class Platform : public ParticleContactGenerator {
 public:
-    Vector3 start;
-    Vector3 end;
-    std::vector<Particle::Ptr>& particles;
-    real const blobRadius;
+    Vector3 const & start;
+    Vector3 const & end;
+    Particles & particles;
+    double blobRadius;
```

Comment:
I could make it const, yes. But I don't see a point of writing trivial setters and getters even in this case.